### PR TITLE
Extractor state machines & base commands

### PR DESCRIPTION
**Pull request by cyyynthia**
Supersedes #9.

Built-in extractors are now written as a combination of a tokenization step powered by TextMate grammar, and a token consumption logic powered by state machines. A statechart representation of the machines in this PR can be viewed here: https://stately.ai/viz/37698a1f-1493-485a-b585-f3017d77fd69

Magic comments are implemented as follows:
 - `@tolgee-ignore` to ignore the next line
 - `@tolgee-key key-name` to specify a key
 - `@tolgee-key { key: 'key-name', ns: 'namespace', defaultValue: '....' }` to specify more than just the key (only key is required) (format doesn't matter, it's parsed as a json5 struct)
 - `@tolgee-key` implies "ignore next statement": it acts as an override

--- 
TM grammars are used in major code editors, which guarantees a high-quality, semantically-correct output. They can then be consumed by a state machine, which has a O(n) time complexity and can be visualized via a statechart. The state machines are powered by xstate, which has a vizualizer tool.

Adding new parsers is as simple as downloading a new grammar via the included script, making the tokenizer aware of the new grammar, and voilà.

### Notes about the download grammar script
Grammars are subject to change as languages mature and new structures are added. The script allows to easily updating all of them by fetching the latest versions of the grammars.

The script includes a transformation routine. This is to allow light modifications to the grammar, which down the road simplifies a *lot* the consumption by the state machine.

### CODE REVIEW

**Comment by cyyynthia** [comment]
Leaving this PR as a draft, as the extractor command itself will need to be updated once the extractor itself is stable

File: src/extractor/tokenizer.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  let linePtr = 0;
+  for (const line of code.split('\n')) {
+    const res = grammar.tokenizeLine(line, stack);
+
+    for (let i = 0; i < res.tokens.length; i++) {
+      const token = res.tokens[i];
+
+      // Opinionated take: if a token is scope-less, chances are we don't care about it.
+      // Ditching it allows us to reduce complexity from the state machine's POV.
+      if (token.scopes.length === 1) continue
```
Comment (stepan662) [inline src/extractor/tokenizer.ts:81]: I would rather use brackets with if statements everywhere. It improves readablity imho.

File: test/extractor/react.test.ts [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    expect(extracted).toEqual(expected);
+  });
+
+  it('overrides the specified namespace if one is passed as parameter', async () => {
+    const expected = [{ keyName: 'key1', namespace: 'ns1' }];
+
+    const code = `
+      function Test () {
+        const { t } = useTranslate('namespace')
+        t('key1', { ns: 'ns1' })
```
Comment (stepan662) [inline test/extractor/react.test.ts:263]: I think you are not covering the case with full object:
```ts
t({ key: 'key', ns: 'namespace', defaultValue: 'default' })
```

**Comment by stepan662** [comment]
Looking good :)

File: test/extractor/react.test.ts (hunk 2 above)
Comment (cyyynthia) [reply test/extractor/react.test.ts:263]: Hm, I don't handle the `t('key', 'default', { ... })` case either... functions with multiple signatures............. 🙃 

I'm a bit bothered by the `key` attribute though, which differs from the `keyName` property the properties state machine is expecting 🤔

Comment (stepan662) [reply test/extractor/react.test.ts:263]: Hmm, the t function needs to be multi signature, as it's the thing that you use all the time and it needs to be comfortable. `keyName` is used on react component as `key` can't be used.

**Comment by JanCizmar** [comment]
I'll leave the CR to Stepan.

**Review by JanCizmar — APPROVED** [review]

File: src/commands/extract/check.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  let warningCount = 0;
+  let filesCount = 0;
+  for (const [file, { warnings }] of extracted) {
+    if (warnings.length) {
+      warningCount += warnings.length;
+      filesCount++;
+
+      const relFile = relative(process.cwd(), file);
+      console.log('%s:', relFile);
```
Comment (JanCizmar) [inline src/commands/extract/check.ts:22]: Calling some wrapped logger function would help us with future enhancing the console output.

File: src/commands/extract/print.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import type { BaseExtractOptions } from '../extract';
+import { relative } from 'path';
+import { Command } from 'commander';
+
+import { extractKeysOfFiles } from '../../extractor';
+import { WarningMessages } from '../../utils/warnings';
+
+type ExtractPrintOptions = BaseExtractOptions;
+
+async function printHandler(this: Command, filesPattern: string) {
```
Comment (JanCizmar) [inline src/commands/extract/print.ts:10]: Is there some reason why to call this function printHandler? and not just print?

File: src/utils/warnings.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    const encodedDescription = description
+      .replaceAll('%', '%25')
+      .replaceAll('\r', '%0D')
+      .replaceAll('\n', '%0A')
+      .replaceAll(':', '%3A')
+      .replaceAll(',', '%2C')
+      .replaceAll('@', '@​'); // This has a ZWS; preventing mentions
+
+    console.log(
+      `::warning file=${file},line=${line},title=${name}::${encodedDescription}`
```
Comment (JanCizmar) [inline src/utils/warnings.ts:65]: COOL!

**Review by JanCizmar — APPROVED** [review]
Seems OK! :)

File: src/commands/extract/print.ts (hunk 4 above)
Comment (cyyynthia) [reply src/commands/extract/print.ts:10]: I just called all the command handlers "somethingHandler", for honestly no specific reason.

**Pull request merged**